frontend/controller/serviceworkers/sw-primary.js

```diff
@@ -45,3 +49,32 @@ self.addEventListener('message', function (event) {
     console.error('[sw] unexpected data:', event.data)
   }
 })
+
+self.addEventListener('push', function (event) {
+  // PushEvent reference: https://developer.mozilla.org/en-US/docs/Web/API/PushEvent
+
+  if (!(self.Notification && self.Notification.permission === 'granted')) {
+    return
+  }
+
+  const data = event.data.json()
+  console.debug('[sw] push received: ', data)
+
+  self.registration.showNotification(
+    data.title,
+    {
+      body: data.body || '',
+      icon: '/assets/images/pwa-icons/group-income-icon-transparent.svg'
+    }
+  )
+})
```

Comment:
What is his the reason [`showNotification`](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerRegistration/showNotification) is being used here instead of `Notification` as used in the [example here](https://developer.mozilla.org/en-US/docs/Web/API/PushEvent#examples)?

Reply:
@taoeffect that was what I found from [this MDN PWA tutorial](https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/Tutorials/js13kGames/Re-engageable_Notifications_Push) (down below in `service-worker.js` section).  But looking at the example link you attached, I think we can benefit more functionalities from using `Notification` instead (e.g. attaching the `click` event listener to the browser notification UI etc.) I will replace it with using `Notification` instead.

Reply:
@SebinSong ok, yeah worth testing it out... let me know if it works that way or not. Maybe it's something special about being inside of a service worker? 🤔